src/compose/types/service.ts

```diff
@@ -28,6 +29,44 @@ export interface ServiceNetworkDictionary {
 	};
 }
 
+// https://docs.docker.com/compose/compose-file/compose-file-v2/#volumes
+export const LongSyntaxVolume = t.intersection([
+	t.type({
+		target: t.string,
+		// https://github.com/gcanti/io-ts/blob/master/index.md#union-of-string-literals
+		type: t.keyof({
+			bind: null,
+			volume: null,
+			tmpfs: null,
+		}),
+	}),
+	t.partial({
+		// long syntax tmpfs disallows specifying 'source', whereas other
+		// long syntax volume types require 'source'
+		source: t.string,
+		readOnly: t.boolean,
+		bind: t.type({
+			propagation: t.keyof({
+				private: null,
+				rprivate: null,
+				shared: null,
+				rshared: null,
+				slave: null,
+				rslave: null,
```

Comment:
This might not be the right place to define this requirement, as we don't validate the syntax when receiving the target state. If someone where to set a wrong value for the `propagation` option, and that makes it all the way to the point where the code is configuring the volumes, then the call `LongSyntaxVolume.is` will return `false`, causing unintended side effects in the comparison.

I think setting `propagation: t.string` should be enough, as any issues will be handled by the engine in that case.

Reply:
docker-compose won't allow an incorrect propagation mode for bind as well, so balena-compose should also handle this. However since we don't allow user-defined bind mounts anyway, there shouldn't be any issue to changing to t.string as suggested, so I've changed this in the latest force push.